config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/tasks/zip/builder/TrustedCertEntityBuilder.java

```diff
@@ -0,0 +1,138 @@
+import static com.ca.apim.gateway.cagatewayconfig.util.xml.DocumentUtils.createElementWithChildren;
+import static com.ca.apim.gateway.cagatewayconfig.util.xml.DocumentUtils.createElementWithTextContent;
+
+public class TrustedCertEntityBuilder implements EntityBuilder {
+    private final Document document;
+    private final IdGenerator idGenerator;
+
+    TrustedCertEntityBuilder(Document document, IdGenerator idGenerator) {
+        this.document = document;
+        this.idGenerator = idGenerator;
+    }
+
+    @Override
+    public List<Entity> build(Bundle bundle) {
+        return bundle.getTrustedCerts().entrySet().stream().map(trustedCertEntry ->
+                buildTrustedCertEntity(trustedCertEntry.getKey(), trustedCertEntry.getValue(), bundle.getCertificateFiles())
+        ).collect(Collectors.toList());
+    }
+
+    private Entity buildTrustedCertEntity(String name, TrustedCert trustedCert, Map<String, String> certificateFiles) {
+        final String id = idGenerator.generate();
+        final Element trustedCertElem = createElementWithAttributesAndChildren(
+                document,
+                TRUSTED_CERT,
+                ImmutableMap.of(ATTRIBUTE_ID, id),
+                createElementWithTextContent(document, NAME, name),
+                buildCertData(trustedCert, certificateFiles)
+        );
+        buildAndAppendPropertiesElement(trustedCert.getProperties(), document, trustedCertElem);
+
+        return new Entity(TRUSTED_CERT_TYPE, name, id, trustedCertElem);
+
+    }
+
+    private Element buildCertData(TrustedCert trustedCert, Map<String, String> certificateFiles) {
+        if (StringUtils.isNotEmpty(trustedCert.getUrl())) {
+            return buildCertDataFromUrl(trustedCert);
+        } else if (StringUtils.isNotEmpty(trustedCert.getFile())) {
+            return buildCertDataFromFile(certificateFiles.get(trustedCert.getFile()));
+        } else if (trustedCert.getCertificateData() != null) {
+            CertificateData certData = trustedCert.getCertificateData();
+            return createElementWithChildren(
+                    document,
+                    CERT_DATA,
+                    createElementWithTextContent(document, ISSUER_NAME, certData.getIssuerName()),
+                    createElementWithTextContent(document, SERIAL_NUMBER, certData.getSerialNumber()),
+                    createElementWithTextContent(document, SUBJECT_NAME, certData.getSubjectName()),
+                    createElementWithTextContent(document, ENCODED, certData.getEncodedData())
+            );
+        } else {
+            throw new EntityBuilderException("Trusted Cert must be loaded from a valid url, from a file, or using certificateData reference.");
+        }
+    }
+
+    private Element buildCertDataFromFile(String certFileLocation) {
+        if (StringUtils.isEmpty(certFileLocation)) {
+            throw new EntityBuilderException("The certificate file location specified does not exist.");
```

Comment:
you may want to say here that the cert file is not specified instead of not found

Reply:
Yes that sounds good